src/TreeSupport.cpp

```diff
@@ -650,32 +610,119 @@ void TreeSupport::insertDroppedNode(std::unordered_set<Node*>& nodes_layer, Node
-        }
+        return it->second;
+    }
+    else
+    {
+        return calculateCollision(key);
     }
 }
 
+const Polygons& ModelVolumes::avoidance(coord_t radius, int layer) const
+{
+    radius = roundRadius(radius);
+    RadiusLayerPair key{radius, layer};
+    const auto it = avoidance_cache_.find(key);
+    if (it != avoidance_cache_.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        return calculateAvoidance(key);
+    }
+}
+
+const Polygons& ModelVolumes::internal_model(coord_t radius, int layer) const
+{
+    radius = roundRadius(radius);
+    RadiusLayerPair key{radius, layer};
+    const auto it = internal_model_cache_.find(key);
+    if (it != internal_model_cache_.end())
+    {
+        return it->second;
+    }
+    else
+    {
+        return calculateInternalModel(key);
+    }
+}
+coord_t ModelVolumes::roundRadius(coord_t radius) const
+{
+    const auto remainder = radius % radius_sample_resolution_;
+    const auto delta = remainder != 0 ? radius_sample_resolution_- remainder : 0;
+    return radius + delta;
+}
+
+const Polygons& ModelVolumes::calculateCollision(const RadiusLayerPair& key) const
+{
+    const auto& radius = key.first;
+    const auto& layer = key.second;
+
+    auto collision_areas = machine_border_;
+    if (layer < static_cast<int>(layer_outlines_.size())) {
+        collision_areas = collision_areas.unionPolygons(layer_outlines_[layer]);
+    }
+    collision_areas = collision_areas.offset(xy_distance_ + radius, ClipperLib::JoinType::jtRound);
+    const auto ret = collision_cache_.insert({key, std::move(collision_areas)});
+    assert(ret.second);
+    return ret.first->second;
+}
+
+const Polygons& ModelVolumes::calculateAvoidance(const RadiusLayerPair& key) const
+{
+    const auto& radius = key.first;
+    const auto& layer = key.second;
+
+    if (layer == 0)
+    {
+        avoidance_cache_[key] = collision(radius, 0);
+        return avoidance_cache_[key];
+    }
+
+    // Avoidance for a given layer depends on all layers beneath it so could have very deep recursion depths if
+    // called at high layer heights
+    constexpr auto max_recursion_depth = 100;
+    if (layer >= max_recursion_depth
+        && avoidance_cache_.find({radius, layer - max_recursion_depth}) == avoidance_cache_.end())
+    {
+        avoidance(radius, layer - max_recursion_depth);
+    }
+    auto avoidance_areas = avoidance(radius, layer - 1).offset(-max_move_).smooth(5);
+    avoidance_areas = avoidance_areas.unionPolygons(collision(radius, layer));
+    const auto ret = avoidance_cache_.insert({key, std::move(avoidance_areas)});
+    assert(ret.second);
+    return ret.first->second;
+}
+
+const Polygons& ModelVolumes::calculateInternalModel(const RadiusLayerPair& key) const
+{
+    const auto& radius = key.first;
+    const auto& layer = key.second;
```

Comment:
Me and my colleague discourage the over-use of `auto`.
On this specific line it's quite unclear what `layer` is - especially because it's not a layer, but a layer *number*.
```suggestion
    const LayerIndex& layer_index = key.second;
```

Reply:
Understood, i'll be the first to admit I can be a bit over enthusiastic with auto.
However, the advantage here is that it makes it trivial to swap out the types (from int to LayerIndex as you recommend above) without having to modify the code because the type interfaces are the same.

I think the issue here is the poor variable naming on my part. I'll clean up the naming, let me know if that provides enough clarity